Reapply #2017 (Allow shapecheck of PixelCNN++), fixing #2245

juliuskunze opened the pull request:
fixes #2245

juliuskunze committed:
Unrevert "Allow shapecheck of PixelCNN++ (google#2017)"

This reverts commit ceab1e3edf1e2395035173dc50f24ce6a27475f6.

juliuskunze committed:
Fix out-of-bound slices (#2245)

juliuskunze committed:
Minor

juliuskunze committed:
Add type annotations

juliuskunze committed:
Merge branch 'master' of https://github.com/google/jax into shapecheck

# Conflicts:
#	jax/numpy/lax_numpy.py

mattjj committed:
Merge remote-tracking branch 'origin/master' into shapecheck

mattjj commented:
Just verified that internal tests and #2245 pass.

I only ran internal tests for JAX itself, not downstream projects; I'm going to run tests for downstream projects now.

mattjj commented:
(Just jotting down some notes as I go through this.)

I noticed a bug in `Poly.__rsub__`, namely that we had written `self + -other` while `self.__rsub__(other)` actually means `other - self`.

mattjj commented on jax/numpy/lax_numpy.py:
```diff
@@ -3316,7 +3321,7 @@ def _index_to_gather(x_shape, idx):
   start_index_map = []
 
-  index_dtype = int64 if _max(x_shape, default=0) >= (1 << 31) else int32
+  index_dtype = int64 if any([type(dim) is Poly or dim >= (1 << 31) for dim in x_shape]) else int32
```
Btw we need to use `_any` not `any` here, since the latter is a `jax.numpy` function! Tricky.

mattjj commented on jax/numpy/lax_numpy.py:
```diff
@@ -3375,7 +3380,8 @@ def _index_to_gather(x_shape, idx):
       i = _normalize_index(i, x_shape[x_axis])
-      i = lax.convert_element_type(i, index_dtype)
+      # dummy index if is polynomial, doesn't matter for shape inference:
+      i = 0 if type(i) is Poly else lax.convert_element_type(i, index_dtype)
```
This seems pretty subtle... hmm...

mattjj replied:
I'm going to leave a todo here to follow up and figure out the best way to do this.

JuliusKunze replied:
`i = lax.convert_element_type(i, index_dtype)` doesn't work for cases like `x[-1]` where `i` becomes a `Poly` representing `n-1`, breaking `convert_element_type`.

Let me know if you can come up with a better solution!

juliuskunze committed:
Fix Poly.__rsub__

juliuskunze committed:
any -> _any

mattjj committed:
tweaks, mostly comments/whitespace

JuliusKunze commented:
Good catch with `__rsub__`! It's now test covered and fixed.

mattjj commented:
Thanks for the quick fix!

I read through in vim and tweaked mostly whitespace/comments in the commit I just pushed. LGTM! Let's wait for Travis to be happy, then I'll merge.

mattjj reviewed (approved):
LGTM! Thanks for this leap forward. It's even better than #2017.

mattjj commented:
Running one more pass of downstream tests to be sure...

mattjj commented:
I see a failure in [this test](https://github.com/tensorflow/probability/blob/e38aab74065a122c556460de6f17f54245ec251b/tensorflow_probability/python/distributions/mvn_tril_test.py#L507) that I must look into.

shoyer commented on jax/interpreters/masking.py:
```diff
@@ -89,31 +86,34 @@ def mask_subtrace(master, in_vals, shape_exprs):
 
-def ensure_poly(p):
-  if isinstance(p, Poly):
+def to_index(x):
+  """Like operator.index, but allowing polymorphic dimensions.
+  Not implemented as `Poly.__index__`, since operator.index only allows ints."""
+  return x if type(x) is Poly else op.index(x)
```
We consolidated similar logic into `canonicalize_shape`/`_canonicalize_dimension` in `core.py`. Could you make use of those here?

(It also helps avoid some potential circular dependency issues.)

JuliusKunze replied:
Ok, nice! I just removed `to_index`, `canonicalize_shape` is now used everywhere.

mattjj commented:
We hit the `_slice_indices` function in that test. If I replace its body with `return idx.indices(size)` then the test passes. Seems like we've got this bug cornered... but it is concerning that this issue didn't turn up in our tests!

mattjj commented:
IIUC `_slice_sizes` has logic based on [this CPython code](https://github.com/python/cpython/blob/37af21b667a9f41437b5b8e451497d7725016df5/Objects/sliceobject.c#L165).

mattjj commented:
I think the riskiest thing with a PR like this is that we break existing code paths, which is unfortunately what this test failure is showing us. A way to avoid that is to be sure to leave the existing paths intact, and only add new clearly-separated paths. For example, instead of adapting the existing `_static_idx` function logic to handle both `Poly` instances and ints in one if/else chain, we can just check up-front whether there are any `Poly`s involved and if so call into a new function; if not, we fall through to the existing logic unchanged.

I'll try to make that change fo `_slice_indices`, but we should go through the PR and look for all opportunities to do that. @JuliusKunze can you help me with that?

mattjj commented:
The issue was that when given these inputs:

```
idx = slice(DeviceArray(44, dtype=int64), DeviceArray(-52, dtype=int64), DeviceArray(-3, dtype=int64))

size = 2
```

We get

```
idx.indices(size) == (1, -1, -3)

_slice_indices(idx, size) ==
(DeviceArray(44, dtype=int64), DeviceArray(-50, dtype=int64), DeviceArray(-3, dtype=int64))
```

mattjj commented:
(I'm a bit concerned that those are DeviceArrays at all... that seems problematic.)

mattjj reviewed (commented):
Actually, before being merge-able we need to
1. make sure we're not affecting existing code paths (as a silent error almost crept through), and
2. ~make sure dimensions are never DeviceArrays.~ <-- this is a problem on master

mattjj commented:
Actually, I think we want to model [this CPython code](https://github.com/python/cpython/blob/6d6508765514c7c10719478a0430f5e47c9a96ac/Objects/sliceobject.c#L372).

shoyer commented:
> 2\. make sure dimensions are never DeviceArrays.

I think we ensure this already, after my change to add coercion via `canonicalize_shape` into the constructor for `ShapedArray`:
https://github.com/google/jax/blob/790d92965c881e0316876242953009ca6fc8f3eb/jax/core.py#L819

mattjj commented:
Perhaps, but not in this code: my comment above shows DeviceArrays present in dimensions in a debugger.

mattjj commented:
By the way, I don't have particular reason to believe that this PR introduced the sizes-as-DeviceArrays issue. It's just when we're noticing it!

mattjj commented:
Ah, it's because of how `_normalize_index` is written and used here. This is an issue on master. I'll just open a separate issue for it.

mattjj committed:
separate polymorphic code path, patch _slice_sizes

mattjj commented:
Okay, I pushed a fix for `_slice_sizes`, basically following that CPython function. I left the Poly case as NotImplemented, which will probably break some tests in this PR. Just wanted to commit what I had so far!

mattjj commented:
@JuliusKunze could you follow up on the Poly cases (where I wrote `raise NotImplementedError  # TODO` lines)? Actually, we also need to handle when `size` is a `Poly` (or perhaps only that case?) so this placeholder logic might not be quite right.

mattjj committed:
put back some logic for handling Poly sizes

mattjj commented:
Actually, I think I got things patched up. @JuliusKunze please take a look and lmk what you think! (How the turntables: now it is you who is the reviewer!)

mattjj committed:
improve test_slice_indices

juliuskunze committed:
Remove to_index, replace with canonicalize_shape

juliuskunze committed:
Fix slicing with polymorphic start/stop

juliuskunze committed:
Test negative step for polymorphic slicing

JuliusKunze commented:
@mattjj Thanks for implementing negative steps for slicing, I added tests to cover this. I am ok with defaulting to the builtin function for the standard case. Added more tests for polymorphic start/stop indices that required slight modifications. Rest LGTM!

juliuskunze committed:
Refactor polymorphic slicing

mattjj commented on jax/api.py:
```diff
@@ -28,6 +28,8 @@
 import inspect
 import itertools as it
+import os
+import string
```
I think these two imports are unused. (Easy to fix later!)

JuliusKunze replied:
Removed!

mattjj commented on jax/numpy/lax_numpy.py:
```diff
@@ -2140,7 +2141,8 @@ def arange(start, stop=None, step=None, dtype=None):
   if stop is None and step is None:
     dtype = dtype or _dtype(start)
-    return lax.iota(dtype, start)  # avoids materializing
+    size = start if type(start) is Poly else lax.convert_element_type(start, onp.uint64)
```
The latter case would make `size` a `DeviceArray`, which seems like something we don't want. (We'd want to use `onp` to control types here.) @JuliusKunze but why do we need this line at all?

JuliusKunze replied:
@mattjj According to the [XLA spec](https://www.tensorflow.org/xla/operation_semantics#iota) `lax.iota` is not supposed to accept `float` size, so I added conversion here and made `iota` more strict. 

In order to minimize the diff, I reverted the change for now and put the conversion to `int` to [where it was before](https://github.com/google/jax/pull/2800/files#diff-08adae3ce53df47932e06971edb68642R1237) (otherwise tests fail since arange accepts float). (`arange`/`iota` currently doesn't work with `shapecheck` anyway. That will be part of the next PR, so no need to check for `Poly` here for now.)

mattjj commented on jax/random.py:
```diff
@@ -277,7 +277,7 @@ def _random_bits(key, bit_width, shape):
     raise TypeError("requesting more random bits than a single call provides.")
 
-  counts = lax.tie_in(key, lax.iota(onp.uint32, max_count))
+  counts = lax.tie_in(key, lax.iota(onp.uint32, max_count.astype(onp.uint32)))
```
Why do we need this change in this PR?

JuliusKunze replied:
We don't (relict from the random implementation taken out of this PR), now removed.

mattjj commented:
Global presubmits look good, but there are a couple unrelated-seeming changes here (commented above) that I'd like to avoid if possible.

juliuskunze committed:
Simplify diff

juliuskunze committed:
Fix shapecheck(iota)

mattjj commented:
Internal tests green!
